pyangbind/lib/serialise.py

```diff
@@ -164,54 +160,271 @@ def default(self, obj, mode="default"):
         )
 
-    def map_pyangbind_type(self, map_val, original_yang_type, obj, mode):
+    def map_pyangbind_type(self, map_val, original_yang_type, obj):
         if map_val in ["pyangbind.lib.yangtypes.RestrictedClass", "RestrictedClassType"]:
             map_val = getattr(obj, "_restricted_class_base")[0]
 
         if map_val in ["pyangbind.lib.yangtypes.ReferencePathType", "ReferencePathType"]:
-            return self.default(obj._get(), mode=mode)
+            return self.default(obj._get())
         elif map_val in ["pyangbind.lib.yangtypes.RestrictedPrecisionDecimal", "RestrictedPrecisionDecimal"]:
-            if mode == "ietf":
-                return six.text_type(obj)
-            return float(obj)
+            # NOTE: this doesn't seem like it needs to be a special case?
+            return self.yangt_decimal(obj)
         elif map_val in ["bitarray.bitarray"]:
             return obj.to01()
         elif map_val in ["unicode"]:
             return six.text_type(obj)
         elif map_val in ["pyangbind.lib.yangtypes.YANGBool"]:
-            if original_yang_type == "empty" and mode == "ietf":
-                if obj:
-                    return [None]
+            if original_yang_type == "empty":
+                # NOTE: previously with IETF mode the code would fall-through if obj was falsey
+                return self.yangt_empty(obj)
             else:
-                if obj:
-                    return True
-                else:
-                    return False
+                return bool(obj)
         elif map_val in ["pyangbind.lib.yangtypes.TypedList"]:
-            return [self.default(i) for i in obj]
-        elif map_val in ["int"]:
-            int_size = getattr(obj, "_restricted_int_size", None)
-            if mode == "ietf" and int_size == 64:
-                return six.text_type(obj)
-            return int(obj)
-        elif map_val in ["long"]:
+            return self.yangt_typed_list(obj)
+        elif map_val in ["int", "long"]:
             int_size = getattr(obj, "_restricted_int_size", None)
-            if mode == "ietf" and int_size == 64:
-                return six.text_type(obj)
-            elif six.PY3:
-                return int(obj)
-            else:
-                return long(obj)
+            return self.yangt_long(obj) if int_size == 64 else self.yangt_int(obj)
         elif map_val in ["container"]:
-            return self._preprocess_element(obj.get(), mode=mode)
+            return self.preprocess_element(obj.get())
         elif map_val in ["decimal.Decimal"]:
-            return six.text_type(obj) if mode == "ietf" else float(obj)
+            return self.yangt_decimal(obj)
+
+    def yangt_int(self, obj):
+        # for values that are 32-bits and under..
+        return int(obj)
+
+    def yangt_long(self, obj):
+        # don't need to sepcial-case for PY3 because we assign `long = int` in the header
+        return long(obj)
+
+    def yangt_identityref(self, obj):
```

Comment:
It's worth adding a comment here that this implementation never gets called with non-IETF serialisation because we have a `string` there and we don't need to modify it. It took me a little to understand that this was the case here.